### ISSUE TITLE
--autofit and -–geometry are only half the size on Retina display.

### ISSUE THREAD

**Issue by Bilalh** [mpv-player/mpv#4218]
### Reproduction steps
Use --autofit=480

### Expected behaviour
The window is 480 pixels wide.

### Actual behaviour
The window is 240 pixels wide.

### Notes 
Work correctly on commit 85c8556 (built on Sat Mar  4 18:34:51 GMT 2017)

### mpv version and platform

OS X 10.11.6 (15G1217) on a MacBook Pro (Retina, Mid 2012)

```
mpv git-076116a (C) 2000-2017 mpv/MPlayer/mplayer2 projects
 built on Fri Mar 10 17:31:43 GMT 2017
ffmpeg library versions:
   libavutil       55.34.101
   libavcodec      57.64.101
   libavformat     57.56.101
   libswscale      4.2.100
   libavfilter     6.65.100
   libswresample   2.3.100
ffmpeg version: 3.2.4
```

### Log file
Running `mpv --no-config --no-osc --no-load-scripts --log-file=output.txt z.webm --autofit=480`

http://sprunge.us/OVEa

**Comment by Akemi** [comment]
i believe you are confusing pixel with points here. if you pass 480 to autofit it opens a window which is exactly 480 pixel wide, that are 240 points.

**Comment by Bilalh** [comment]
Is there a way to get it to use points then? It would be annoying to have to use a different command when using a Retina/non-Retina display.

If you use command cmd-shift-4 on the window it says 240 for the width, which caused my confusion on the unit.

**Comment by Akemi** [comment]
no, atm there is no way to use points. geometry and autofit also explicitly state pixel in the docs. so as it stands now the implementation is right from the description, but might be unexpected in the case mentioned.

i would need @wm4's input if this behaviour is wanted and a 'wont fix' or if it should be changed.

specifically the question is, should the geometry and the autofit options consider the `hidpi-window-scale` option when absolute values are used (pixel) or is it an exact pixel value (not points)? as an example `--autofit=480 --hidpi-window-scale=yes`, should the resulting window on a HiDPI resolution (2x scaled) be 480px/240pt wide or 960px/480pt.

**Comment by Bilalh** [comment]
I think it should 480pt with -hidpi-window-scale=yes (which I did not know is on by default on OS X)  since it was previous behaviour and it make more sense to me.

**Comment by Akemi** [comment]
first of all, i am not disagreeing with that from a macOS point of view. also for reference it was broken by the fix for #4194, so we can't easily revert to the previous code.

i just discussed this with wm4. it's basically a matter of forcing the point paradigm of macOS onto all the other platforms or keeping the current behaviour and fixing that problem in the macOS part of the code. we decided to go with the latter.

**Comment by sCreami** [comment]
> it was broken by the fix for #4194

Sorry :kissing: but why not go the webdev way of handling this and allow to set autofit as 480pt or 480px? It would be a multi-platforms solution.

**Comment by Akemi** [comment]
i didn't decide how to fix this yet. also adding an option to use 'pt' alongside of 'px' for those options just means additional work besides fixing the fundamental problem that needs to be handled anyway.

**Comment by zmwangx** [comment]
> i believe you are confusing pixel with points here. if you pass 480 to autofit it opens a window which is exactly 480 pixel wide, that are 240 points.

> geometry and autofit also explicitly state pixel in the docs. so as it stands now the implementation is right from the description, but might be unexpected in the case mentioned.

That's not true. Scaled resolutions on macOS is not always 2x; there are five scaled settings in System Preferences on a 15'' Retina MBP. For instance, when I scale the 2880x1800 physical display to 1920x1200, I have to specify `--geometry=3840x2400` to fill the screen. There are not even that many pixels. See #4357 (a duplicate report) for details.

**Comment by Akemi** [comment]
i am not talking about the virtual screen resolution, it's about the scaling of UI elements. it is always 2 times on HiDPI setups.

from the docs of the [backingScaleFactor](https://developer.apple.com/reference/appkit/nswindow/1419459-backingscalefactor)

> The value of this property is 2.0 for high-resolution scaled display modes, and 1.0 for all other cases.

you even said it yourself. you're setting your physical 2880x1800 resolution so that it appears like a 1920x1200 one. so you have to set geometry to 3840x2400 to fill your screen. that is twice the resolution of your 1920x1200, eg backingScaleFactor = 2.0.

**Comment by zmwangx** [comment]
Well, as I quoted

> geometry and autofit also explicitly state pixel in the docs. so as it stands now the implementation is right from the description, but might be unexpected in the case mentioned.

Either the implementation is not right from the description, or your definition of "pixel" (I read the docs you linked to, and I get that) is not meaningful to real world users, as shown by my 3840x2400 example. I can certainly modify my scripts to account for that, but IMO the old behavior is more reasonable.

**Comment by Akemi** [comment]
in your example the virtual screen resolution is 3840x2400px and you have to set geometry to 3840x2400px to fill the entire screen. hence the description is right. anything else are points.

you even quoted me and the crucial part:

> but might be unexpected in the case mentioned.

yes it's not meaningful and i never claimed otherwise. even your example is not meaningful for "real world users", since they don't see any resolutions at all just "more space", "larger text" etc.

no the old behaviour is less reasonable since you can't meaningfully use percentages to scale the window size, but you still can compensate for the current odd behaviour like you yourself said.

to repeat myself, i haven't decided yet how to fix this properly without breaking current use cases or the calculations on other platforms. i also don't want too many redundancies and there are also some questions how everything is supposed to work in combination with the `--hidpi-window-scale` option.

**Comment by zmwangx** [comment]
One final comment:

> no the old behaviour is less reasonable since you can't meaningfully use percentages to scale the window size

I don't think `--geometry=1920x1200` to fill the screen is at odds with `--geometry=100%x100%` fill the screen. I never said the percentage behavior as outlined in #4194 is reasonable; I was only talking about pixels.

**Comment by Bilalh** [comment]
The old  behaviour is _more_ reasonable if you have multiples monitor which are a mixture of HIPI and non-HIPI.  It means you can have one command e.g. `--autofit=480` which make a window the same number of points on each screen.  I actually when back to using 85c8556 (the commit before the change)  because the new behaviour is too inconvenient.

**Comment by sCreami** [comment]
> He who rejects change is the architect of decay.
― Harold Wilson

**Comment by Akemi** [comment]
you didn't mention the percentage directly but it seems like you are arguing to revert the commit that introduced the current issue. that would mean the percentage case would break again, hence why i mentioned it.

if you didn't argue for reverting that commit then i wonder why you argue at all. i already acknowledged the issue and that i will eventually fix it.

**Comment by Bilalh** [comment]
Sorry, I misread the comment `old behaviour is less reasonable` .  I  mistakenly thought you changed your mind.

**Comment by bodayw** [comment]
I have just realized that on recent MacBook models (namely all 12-inch MacBook, MacBook Pro 2016 and later), Apple now set the display to a "scaled" setting by default. For example, on a 15-inch MacBook Pro the default is now 3300x2100 (looks like 1680x1050), instead of the pixel perfect 2880x1800.

More usable screen space aside, for video playback this is obviously suboptimal. Video will be rendered to a resolution higher than the physical resolution of the screen and then scaled back, which is not only less efficient, but also might introduce unnecessary scaling artifacts.

Is there a way that mpv could circumvent this without changing the default system display setting on those Macs? It seems unlikely judging from what I've read, just want to make sure.

Sorry for hijacking the thread, but I think this is somehow related.

**Comment by Argon-** [comment]
No. mpv is doing it exactly as it should (according to Apple) and there's no official way around rendering to bigger-than-necessary resolutions (and afaik no unofficial ones either).
There is no way around this for you except changing the resolution of your screen to a non-native one (one with scale factor 1, e.g. using https://github.com/avibrazil/RDM) but this also brings the known disadvantages of using non-native resolutions on LCDs along with it.

**Comment by bodayw** [comment]
Oh yes I do agree mpv is doing it right. It's just Apple's decision is a bit ridiculous to me, and thinking whether it's possible to save people who are not aware of resolution and stuff from this issue.

>There is no way around this for you except changing the resolution of your screen to a non-native one (one with scale factor 1, e.g. using https://github.com/avibrazil/RDM) but this also brings the known disadvantages of using non-native resolutions on LCDs along with it.

This confused me. Apple's default setting on those Macs has the exact said known disadvantages of using non-native resolutions, right? To avoid that, one just need to set the resolution of their Macs to the native one, i.e. 2880x1800 (looks like 1440x900) on a 15-inch MacBook Pro, and 2560x1600 (looks like 1280x800) on a 13-inch MacBook Pro etc.

**Comment by Argon-** [comment]
Ah yeah you can do that but it won't help much as you still render to huge resolutions. 
No wht I meant was to select 1440x900 as resolution.

**Comment by abrahamneben** [comment]
This is really quite bothersome. I really want to specify `mpv --geometry=200pt x 400pt` so my video can appear the same size on both retina and non-retina displays, both mac and linux. As an alternative, does anyone know a clean, cross-platform way of getting the points-to-pixels conversion so I can convert manually?

**Comment by Argon-** [comment]
> cross-platform way of getting the points-to-pixels conversion so I can convert manually?

Not sure if such a thing exists. Even worse, on OSX I'm not aware of a single onboard/native program that allows you to at least shell out to for querying this ratio. That's why I have my own ~5 line program doing only that. Not a huge issue but annoying.

Could mpv maybe provide this value as a property? For OSX, it's queried/used anyway: https://github.com/mpv-player/mpv/search?utf8=%E2%9C%93&q=backingScaleFactor

**Comment by Akemi** [comment]
> Could mpv maybe provide this value as a property (as in: is it willing to)?

from my side there wouldn't be a problem with this. it could be done similar to the [minimise querying](https://github.com/mpv-player/mpv/blob/master/video/out/cocoa_cb_common.swift#L426) via VOCTRLs. though Apple [discourages](https://developer.apple.com/documentation/appkit/nsscreen/1388385-backingscalefactor) the use of it somewhat, but something like this probably falls within those "rare cases".

**Comment by abrahamneben** [comment]
If you interact with MPV windows in python or in shell scripts, you can get the number of pixels per point like this, then manually convert the values you give to `--geometry`

$ pip install pyobjc-framework-cocoa
$ python -c "import Cocoa; print(Cocoa.NSScreen.mainScreen().backingScaleFactor())
2.0

**Comment by Akemi** [comment]
that won't work on multi monitor setups since you want to get the backing scale factor of the monitor the mpv window is on and not the the main screen. so you have to get the NSWindow from mpv somehow.
